canary/producer-cpp/jobs/orchestrator.groovy

```diff
@@ -0,0 +1,214 @@
+    return Jenkins.instance
+                    .getAllItems(Job.class)
+                    .findAll(filterClosure)
+}
+
+NEXT_AVAILABLE_RUNNER = null
+ACTIVE_RUNNERS = [] 
+
+pipeline {
+    agent {
+        label 'producer-uw2'
+    }
+
+    options {
+        disableConcurrentBuilds()
+    }
+
+    stages {
+        stage('Checkout') {
+            steps {
+                checkout([$class: 'GitSCM', branches: [[name: GIT_HASH ]],
+                          userRemoteConfigs: [[url: GIT_URL]]])
+            }
+        }
+
+        stage('Update runners') {
+            stages {
+                stage("Find the next available runner and current active runners") {
+                    steps {
+                        script {
+                            def runners = findRunners()
+                            def nextRunner = null 
+                            def oldestTimestamp = Long.MAX_VALUE
+
+                            // find the least active runner
+                            runners.each {
+                                def timestamp = getJobLastBuildTimestamp(it)
+                                if ((it.isDisabled() || !it.isBuilding()) && timestamp < oldestTimestamp) {
+                                    nextRunner = it
+                                    oldestTimestamp = timestamp
+                                }
+                            }
+
+                            if (nextRunner == null) {
+                                error "There's no available runner"
+                            }
+
+                            NEXT_AVAILABLE_RUNNER = nextRunner.getDisplayName()
+                            echo "Found next available runner: ${NEXT_AVAILABLE_RUNNER}"
+
+                            ACTIVE_RUNNERS = runners.findAll({ item -> item != nextRunner && (!item.isDisabled() || item.isBuilding()) })
+                                                    .collect({ item -> item.getDisplayName() })
+                            echo "Found current active runners: ${ACTIVE_RUNNERS}"
+                        }
+                    }
+                }
+            
+                stage("Spawn new runners") {
+                    steps {
+                        script {
+                            echo "New runner: ${NEXT_AVAILABLE_RUNNER}"
+                            Jenkins.instance.getItemByFullName(NEXT_AVAILABLE_RUNNER).setDisabled(false)
+
+                            def gitHash = sh(returnStdout: true, script: 'git rev-parse HEAD')
+                            COMMON_PARAMS << string(name: 'GIT_HASH', value: gitHash)
+                        }
+
+                        // TODO: Use matrix to provide configurability in parameters
+                        build(
+                            job: NEXT_AVAILABLE_RUNNER,
+                            parameters: COMMON_PARAMS + [
+                                string(name: 'CANARY_STREAM_NAME', value: "Continuous-Longrun-01"),
```

Comment:
As epr our discussion, can we reduce the number of builds to unique runs? As per this, we would only need a Realtime long run and realtime periodic. Also change the way the stream names are configured. Remove the `-01`, `-02`, etc.